File: backend/app/services/motor_a.py
from decimal import Decimal
from typing import Any

PARAMS = {"BENCHMARK_RESERVA_MESES": 3}


def calcular_motor_a(
    ahorro: float,
    rentas: float,
    otros: float,
    gastos_basicos: float,
    obligaciones: float,
    creditos: float,
    liquidez: float | None = None,
) -> dict[str, Any]:
    # ─── FIX-1: Ingresos Totales ────────────────────────────────────────────
    # El Excel trata `ahorro` como la CAPACIDAD DE AHORRO NETA ya neteada de
    # rentas y otros ingresos. El ingreso laboral se reconstruye sustrayendo
    # rentas y otros del ahorro:
    #   ingreso_laboral = gastos_basicos + obligaciones + creditos + ahorro − rentas − otros
    #   ingresos_totales = ingreso_laboral + rentas + otros
    #                    = gastos_basicos + obligaciones + creditos + ahorro
    # Verificación Juan Pérez:
    #   ingreso_laboral = (40K+20K+0+50K) − 10K − 0 = $100,000  ← Excel C16 ✓
    #   ingresos_totales = 100K + 10K + 0 = $110,000              ← Excel total ✓
    ingreso_laboral = (
        Decimal(str(gastos_basicos)) + Decimal(str(obligaciones)) + Decimal(str(creditos))
        + Decimal(str(ahorro)) - Decimal(str(rentas)) - Decimal(str(otros))
    )
    ingresos_totales = ingreso_laboral + Decimal(str(rentas)) + Decimal(str(otros))

    gastos_totales = (
        Decimal(str(gastos_basicos)) + Decimal(str(obligaciones)) + Decimal(str(creditos))
    )
    ingresos_val = float(ingresos_totales)

    # Distribución coherente (los 4 componentes suman 100% del ingreso total)
    distribucion = {
        "gastos_pct":      float(gastos_basicos / ingresos_val) if ingresos_val > 0 else 0,
        "obligaciones_pct": float(obligaciones   / ingresos_val) if ingresos_val > 0 else 0,
        "creditos_pct":    float(creditos        / ingresos_val) if ingresos_val > 0 else 0,
        "ahorro_pct":      float(ahorro          / ingresos_val) if ingresos_val > 0 else 0,
    }

    # ─── Reserva de Emergencia ──────────────────────────────────────────────
    benchmark_reserva = PARAMS["BENCHMARK_RESERVA_MESES"] * gastos_basicos
    meses_cubiertos = None
    if liquidez is not None and gastos_basicos > 0:
        meses_cubiertos = liquidez / gastos_basicos

    resultado_reserva = "Pendiente"
    if meses_cubiertos is not None:
        resultado_reserva = "Excedida" if meses_cubiertos >= 3 else "Insuficiente"

    # ─── FIX ADD-1: Remanente real para objetivos ───────────────────────────
    # Si la reserva está cubierta, el remanente = ahorro completo.
    # Si hay déficit de reserva, el remanente se reduce por lo que falta cubrir.
    pendiente_reserva = 0.0
    meses_para_cubrir = 0.0
    if liquidez is not None and meses_cubiertos is not None and meses_cubiertos < 3:
        pendiente_reserva = benchmark_reserva - (liquidez or 0)
        meses_para_cubrir = (pendiente_reserva / ahorro) if ahorro > 0 else 0.0

    remanente = max(0.0, ahorro * (1 - min(meses_para_cubrir / 12, 1)) if meses_para_cubrir > 0 else ahorro)

    return {
        "ingresos_totales":   float(ingresos_totales),
        "gastos_totales":     float(gastos_totales),
        "distribucion":       distribucion,
        "benchmark_reserva":  benchmark_reserva,
        "meses_cubiertos":    meses_cubiertos,
        "resultado_reserva":  resultado_reserva,
        "pendiente_reserva":  pendiente_reserva,
        "meses_para_cubrir":  round(meses_para_cubrir, 2),
        "remanente":          remanente,
    }

File: src/lib/motors/motor-a.ts
import { PARAMS } from "../constants";

export interface MotorAInput {
  ahorro: number;
  rentas: number;
  otros: number;
  gastos_basicos: number;
  obligaciones: number;
  creditos: number;
  liquidez?: number;
}

export interface MotorAOutput {
  ingresos_totales: number;
  gastos_totales: number;
  distribucion: {
    gastos_pct: number;
    obligaciones_pct: number;
    creditos_pct: number;
    ahorro_pct: number;
  };
  benchmark_reserva: number;
  meses_cubiertos: number | null;
  resultado_reserva: "Excedida" | "Insuficiente" | "Pendiente";
  pendiente_reserva: number;
  meses_para_cubrir: number;
  remanente: number;
  // Desglose activo vs pasivo de la capacidad de ahorro (Excel fuente de verdad)
  // ingreso_disponible_pasivo = min(rentas + otros, ahorro)  — de cuánto ahorro viene de ingresos pasivos
  // ingreso_disponible_activo = ahorro − ingreso_disponible_pasivo — lo que viene de trabajo activo
  ingreso_disponible_pasivo: number;
  ingreso_disponible_activo: number;
}

export function calcularMotorA(input: MotorAInput): MotorAOutput {
  // ─── Ingresos Totales ───────────────────────────────────────────────────
  // Ingresos totales = gastos_basicos + obligaciones + créditos + ahorro mensual
  // Representa todo lo que el cliente percibe y distribuye cada mes.
  const ingresos_totales =
    input.gastos_basicos + input.obligaciones + input.creditos + input.ahorro;
  const gastos_totales =
    input.gastos_basicos + input.obligaciones + input.creditos;

  // Distribución coherente: los 4 componentes suman 100% del ingreso total
  const distribucion = {
    gastos_pct:       ingresos_totales > 0 ? input.gastos_basicos / ingresos_totales : 0,
    obligaciones_pct: ingresos_totales > 0 ? input.obligaciones   / ingresos_totales : 0,
    creditos_pct:     ingresos_totales > 0 ? input.creditos        / ingresos_totales : 0,
    ahorro_pct:       ingresos_totales > 0 ? input.ahorro          / ingresos_totales : 0,
  };

  // ─── Reserva de Emergencia ──────────────────────────────────────────────
  const benchmark_reserva = PARAMS.BENCHMARK_RESERVA_MESES * input.gastos_basicos;

  let meses_cubiertos: number | null = null;
  if (input.liquidez !== undefined && input.gastos_basicos > 0) {
    meses_cubiertos = input.liquidez / input.gastos_basicos;
  }

  let resultado_reserva: "Excedida" | "Insuficiente" | "Pendiente" = "Pendiente";
  if (meses_cubiertos !== null) {
    resultado_reserva = meses_cubiertos >= 3 ? "Excedida" : "Insuficiente";
  }

  // ─── FIX ADD-1: Remanente real para objetivos ───────────────────────────
  // Si hay déficit en reserva, parte del ahorro mensual se destina a cubrirlo.
  let pendiente_reserva = 0;
  let meses_para_cubrir = 0;
  if (meses_cubiertos !== null && meses_cubiertos < 3 && input.liquidez !== undefined) {
    pendiente_reserva = benchmark_reserva - (input.liquidez ?? 0);
    meses_para_cubrir = input.ahorro > 0 ? pendiente_reserva / input.ahorro : 0;
  }

  const remanente =
    meses_para_cubrir > 0
      ? Math.max(0, input.ahorro * (1 - Math.min(meses_para_cubrir / 12, 1)))
      : input.ahorro;

  // ─── Desglose ingreso disponible activo vs pasivo ────────────────────────
  // Ingreso pasivo disponible: de tu capacidad de ahorro, cuánto proviene de
  // fuentes pasivas (rentas + negocios). Tope: no puede superar el ahorro total.
  // Ingreso activo disponible: lo que resta, proveniente de tu trabajo activo.
  const ingreso_disponible_pasivo = Math.min(input.rentas + input.otros, input.ahorro);
  const ingreso_disponible_activo = Math.max(0, input.ahorro - ingreso_disponible_pasivo);

  return {
    ingresos_totales,
    gastos_totales,
    distribucion,
    benchmark_reserva,
    meses_cubiertos,
    resultado_reserva,
    pendiente_reserva,
    meses_para_cubrir: Math.round(meses_para_cubrir * 100) / 100,
    remanente,
    ingreso_disponible_pasivo: Math.round(ingreso_disponible_pasivo * 100) / 100,
    ingreso_disponible_activo: Math.round(ingreso_disponible_activo * 100) / 100,
  };
}
